### PR TITLE
DataChannel bufferedamount

### PR DESCRIPTION
**Pull request by jmillan**
* DataConsumer.getBufferedAmount() method
* DataConsumer.bufferedAmountLowThreshold
* DataConsumer 'bufferedamountlow' event

### CODE REVIEW

File: worker/src/RTC/SctpAssociation.cpp [hunk 1]
```diff
@@ -406,6 +447,9 @@ namespace RTC
 			  ppid,
 			  len,
 			  std::strerror(errno));
+
+			// Decrease buffered amount.
+			SetSctpBufferedAmount(this->sctpBufferedAmount - len);
```
Comment (ibc) [inline worker/src/RTC/SctpAssociation.cpp]: I don't like this as per our conversation. Just set and emit if ret is 0.

File: worker/src/RTC/Transport.cpp [hunk 2]
```diff
@@ -2788,6 +2805,24 @@ namespace RTC
...
+		MS_TRACE();
+
+		// Notify the Node Transport.
+		json data = json::object();
+
+		data["bufferedAmount"] = len;
+
+		for (const auto& kv : this->mapDataConsumers)
+		{
+			const auto* dataConsumer = kv.second;
```
Comment (ibc) [inline worker/src/RTC/Transport.cpp:2828]: This should be just for SCTP type DataConsumers (and not for direct ones).

**Review by ibc — CHANGES_REQUESTED** [review]
Looks good, just a few issues.

**Review by ibc — APPROVED** [review]

**Review by ibc — APPROVED** [review]

File: src/DataConsumer.ts [hunk 3]
```diff
@@ -148,6 +155,19 @@ export class DataConsumer extends EnhancedEventEmitter
...
 
+	/**
+	 * Buffered amount threshold.
+	 */
+	get bufferedAmountLowThreshold(): number
+	{
+		return this._bufferedAmountLowThreshold;
+	}
+
+	set bufferedAmountLowThreshold(value: number)
```
Comment (ibc) [inline src/DataConsumer.ts:166]:
```suggestion
	set bufferedAmountLowThreshold(value: number): void
```

Comment (jmillan) [reply src/DataConsumer.ts:166]: A set accessor cannot have a return type annotation.

File: src/DataConsumer.ts [hunk 4]
```diff
@@ -316,9 +349,29 @@ export class DataConsumer extends EnhancedEventEmitter
...
+					const { bufferedAmount } = data as { bufferedAmount: number };
+
+					if (
+						bufferedAmount <= this._bufferedAmountLowThreshold &&
+						bufferedAmount <= this._bufferedAmount
+					)
+					{
+						this.safeEmit('bufferedamountlow', bufferedAmount);
+
+						// Emit observer event.
```
Comment (ibc) [inline src/DataConsumer.ts]: Let's not emit this event in the observer, please. Anyway it's not documented within the `observer` inline doc, but better do not emit this in observer.

Comment (jmillan) [reply src/DataConsumer.ts]: done

File: src/DataConsumer.ts [hunk 5]
```diff
@@ -316,9 +349,29 @@ export class DataConsumer extends EnhancedEventEmitter
...
 				}
 
+				case 'bufferedamount':
+				{
+					const { bufferedAmount } = data as { bufferedAmount: number };
+
+					if (
+						bufferedAmount <= this._bufferedAmountLowThreshold &&
+						bufferedAmount <= this._bufferedAmount
+					)
```
Comment (ibc) [inline src/DataConsumer.ts]: This code is wrong. It does not work. The condition must be:

- New buffered amount is less or equal than threshold and current buffered amount is higher than threshold.

Also, please update `this._bufferedAmount` before emitting events:

```ts
const { bufferedAmount } = data as { bufferedAmount: number };
const previousBufferedAmount = this._bufferedAmount;

this._bufferedAmount = bufferedAmount;

if (
  previousBufferedAmount > this._bufferedAmountLowThreshold &&
  this._bufferedAmount <= this._bufferedAmountLowThreshold
)
{
  this.safeEmit('bufferedamountlow', this._bufferedAmount);
}
```

Comment (ibc) [reply src/DataConsumer.ts]: Also remove the 'sctpsendbufferfull' from the observer, please.

Comment (jmillan) [reply src/DataConsumer.ts]: alright

Comment (ibc) [reply src/DataConsumer.ts]: This is marked as resolved, but `sctpsendbufferfull` is still emitted by the observer.

Comment (jmillan) [reply src/DataConsumer.ts]: ah, I didn't read the `sctpsendbufferfull` comment.

Comment (ibc) [reply src/DataConsumer.ts]: This must also be removed from the docs: https://mediasoup.org/documentation/v3/mediasoup/api/#dataConsumer-observer-on-sctpsendbufferfull

**Comment by jmillan** [comment]
> This must also be removed from the docs: https://mediasoup.org/documentation/v3/mediasoup/api/#dataConsumer-observer-on-sctpsendbufferfull

Already done

**Comment by jmillan** [comment]
deploying docs...

**Comment by ibc** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- lib/DataConsumer.js  3
         

Clones added
============
- lib/DataConsumer.js  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/versatica/mediasoup/pullRequest?prid=5955719&bid=19519240)

**Pull request merged**